BambooTracker/tick_counter.cpp

```diff
@@ -130,19 +139,51 @@ void TickCounter::resetCount()
 {
 	restTickToNextStep_ = 0;
 	tickDiffSum_ = 0;
+	prevTickDiffSum_ = 0;
 }
 
+// MUST be idempotent when no groove is active.
 void TickCounter::resetRest()
 {
 	if (nextGroovePos_ == -1) {
-		tickDiffSum_ += tickDiff_;
+		tickDiffSum_ = prevTickDiffSum_ + tickDiff_;
 		int castedTickDifSum = static_cast<int>(tickDiffSum_);
 		restTickToNextStep_ = defStepSize_ + castedTickDifSum;
-		if (restTickToNextStep_ < 1) restTickToNextStep_ = 1;	// Prevent wait count changing to 0 (freeze)
 		tickDiffSum_ -= castedTickDifSum;
 	}
 	else {
 		restTickToNextStep_ = grooves_.at(static_cast<size_t>(nextGroovePos_));
 		nextGroovePos_ = (nextGroovePos_ + 1) % static_cast<int>(grooves_.size());
 	}
+
+	// restTickToNextStep_ can be 0 in case of a groove containing 0.
+	// Previously this function wrote 0, and TickCounter::countUp() subtracted 1,
+	// resulting in restTickToNextStep_ = -1, and waiting 4 billion ticks
+	// (plus signed integer overflow UB) for it to reach 0 and advance to the next row.
+	// To avoid this, clamp non-positive values to 1 in both speed and groove.
+	if (restTickToNextStep_ < 1) restTickToNextStep_ = 1;
+
+	/*
+	There are several cases when resetRest() can be called:
+
+	- When loading file (resetRest() must not decrement restTickToNextStep_)
+	- When retrieving speed/tempo/groove effects (restTickToNextStep_ doesn't matter,
+	  will be overwritten by countUp())
+	- After countUp() processes head of step (resetRest() must decrement restTickToNextStep_)
+	- When changing a song's global speed, tempo, or groove, including when playback is
+	  active (the exact behavior is unimportant)
+
+	Currently, resetRest() decrements restTickToNextStep_ if the song is playing.
+
+	There are other cases when resetRest() is never called (I think) unless you change
+	global settings:
+
+	- Before countUp() processes head of step (the current code always calls countUp()
+	  before calling resetRest())
+	- After countUp() processes middle of step (never happens, Fxx effects are always
+	  processed at the head, not delayed by a Gxx effect)
+	*/
```

Comment:
This whole comment feels very excessive for a simple `if (something == true) then decrement(thatThing) end`. Some of that might be more appropriate in a commit description or function-level comment?